core/nightshadev2/src/nightshade.rs

```diff
@@ -0,0 +1,370 @@
+use std::cmp::Ordering;
+use std::collections::HashSet;
+
+pub type AuthorityId = usize;
+pub type BLSSignature = u64;
+
+const COMMIT_THRESHOLD: i64 = 3;
+
+pub enum NSResult {
+    Updated(Option<State>),
+    Error(String),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct BareState {
+    endorses: AuthorityId,
+    confidence0: i64,
+    confidence1: i64,
+}
+
+impl PartialOrd for BareState {
+    fn partial_cmp(&self, other: &BareState) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for BareState {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.confidence0 != other.confidence0 {
+            return if self.confidence0 > other.confidence0 {
+                Ordering::Greater
+            } else {
+                Ordering::Less
+            };
+        }
+
+        if self.endorses != other.endorses {
+            return if self.endorses < other.endorses {
+                Ordering::Greater
+            } else {
+                Ordering::Less
+            };
+        }
+
+        if self.confidence1 != other.confidence1 {
+            return if self.confidence1 > other.confidence1 {
+                Ordering::Greater
+            } else {
+                Ordering::Less
+            };
+        }
+
+        Ordering::Equal
+    }
+}
+
+impl BareState {
+    fn new(endorses: AuthorityId) -> Self {
+        Self {
+            endorses,
+            confidence0: 0,
+            confidence1: 0,
+        }
+    }
+
+    fn empty() -> Self{
+        Self{
+            endorses: 0,
+            confidence0: -1,
+            confidence1: -1,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct SignedState {
+    signature: BLSSignature,
+    parent: Vec<BareState>,
+}
+
+impl SignedState {
+    fn new() -> Self {
+        Self {
+            signature: 0,
+            parent: vec![],
+        }
+    }
+
+    fn update(&mut self, state: &State) {
+        // TODO: Update self.signature using state.get_signature
+        self.parent.push(state.bare_state.clone());
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
```

Comment:
don't derive equality, we should only compare BareStates

Reply:
[Fixed](https://github.com/nearprotocol/nearcore/pull/549/commits/67b71d2feaf245696c38cc5b027521d22e87811c)